src/OrchardCore.Modules/OrchardCore.Layers/Indexes/LayerMetadataIndex.cs

```diff
@@ -14,19 +14,20 @@ public class LayerMetadataIndexProvider : IndexProvider<ContentItem>
         public override void Describe(DescribeContext<ContentItem> context)
         {
             context.For<LayerMetadataIndex>()
-                .When(contentItem => contentItem.Has<LayerMetadata>())
+                // Keep index records of soft deleted items as they are contained items.
+                .When(contentItem => contentItem.Has<LayerMetadata>() && (contentItem.Published || contentItem.Latest))
```

Comment:
Hmm, ok, so I like the `.When(contentItem => contentItem.Has<LayerMetadata>()` but I don't think we're gaining anything with `&& (contentItem.Published || contentItem.Latest)` - it's a micro optimization I think, which makes the api more complicated (and potentially prone to strange behaviour)

Say for example you have a handler that catches and changes something about this, when the document is unpublished, then it doesn't get saved to the index.

I'm not dead against it, but it seems like a micro optimization, that just makes the api more confusing?

Reply:
@deanmarcussen 

Yes i agree about optimization but i only thought about what i should do, keep the index when removing an item (as it is a contained item) that was already indexed as an active(s) version(s), but still not indexing all versions, e.g. when we publish / draft / publish / draft, and so on.

Just re-tried as it was before, indeed we enlist all versions, there is not so many items in a layer, but more blog posts. Is it what we want? If so i'm fine with it.

Then i re-tried as it is in this PR, we index only active versions, but when we delete an item, and i was surprised, the index is not kept in the table, so yes, at least moving the check from the mapping to the `When()` is useless.

But this is because **when an item is filtered by the `When()`, it is still removed from the index**, this is what i thought initially (i may have tested it) but, as i remember, for another index table you said to me the opposite ;) So we removed a filter in the `.When()` as we thought that it would not remove the index row as it is suposed to do.

But not sure at 100%, i will redo tests tonight, i may be wrong. Then i will check if we can still keep the last actives versions when removing an item version (e.g. unpublish a not latest one), but without indexing all the history of all versions, or just let it as it was before for the 2 related index tables ;)

Reply:
> But this is because when an item is filtered by the When(), it is still removed from the index,

This shouldn't happen. `When` should only be an indicator to YesSql that it should call `Map`. When it returns true, it should call `Map` which will force the delete and when it returns null, will not add the creates. But remember it's only being called on the content item being saved, not reindexing the whole index table.

So I would expect that say, for example, publishing an item, results in one call to reset the indexes on the old item (i.e. delete and recreate), and one call for the new item. Because only those two items are being saved.